go/store/blobstore/oci.go

```diff
@@ -0,0 +1,796 @@
+		UploadId:       &uploadID,
+		UploadPartNum:  &partNumber,
+		ContentLength:  &contentLength,
+		UploadPartBody: io.NopCloser(reader),
+	})
+	if err != nil {
+		return objectstorage.CommitMultipartUploadPartDetails{}, err
+	}
+
+	return objectstorage.CommitMultipartUploadPartDetails{
+		Etag:    res.ETag,
+		PartNum: &partNumber,
+	}, nil
+}
+
+func (bs *OCIBlobstore) downloadPart(ctx context.Context, objectName, namespace, etag string, partNumber int, br BlobRange) (*toDownload, error) {
+	if objectName == "" {
+		return nil, errors.New("object name required to upload part")
+	}
+
+	req := objectstorage.GetObjectRequest{
+		NamespaceName: &namespace,
+		BucketName:    &bs.bucketName,
+		ObjectName:    &objectName,
+	}
+
+	star := "*"
+	if etag != "" {
+		req.IfMatch = &etag
+	} else {
+		req.IfNoneMatch = &star
+	}
+
+	byteRange := br.asHttpRangeHeader()
+	if byteRange != "" {
+		req.Range = &byteRange
+	}
+
+	res, err := bs.client.GetObject(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer res.Content.Close()
+
+	p, err := createTempFile()
+	if err != nil {
+		return nil, err
+	}
+
+	w, err := os.OpenFile(p, os.O_APPEND|os.O_WRONLY, os.ModePerm)
+	if err != nil {
+		return nil, err
+	}
+	defer w.Close()
+
+	_, err = io.Copy(w, res.Content)
+	if err != nil {
+		return nil, err
+	}
+
+	return &toDownload{
+		path:    p,
+		partNum: partNumber,
+		br:      br,
+	}, nil
+}
+
+func createTempFile() (string, error) {
+	f, err := os.CreateTemp("", "")
+	if err != nil {
+		return "", err
+	}
+	defer f.Close()
+	return f.Name(), nil
+}
+
+func copyFileToWriter(path string, w io.Writer) error {
+	r, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer r.Close()
+	_, err = io.Copy(w, r)
+	return err
+}
+
+func appendPartsToFile(path string, completedParts []*toDownload) error {
+	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	for _, p := range completedParts {
+		err = copyFileToWriter(p.path, f)
+		if err != nil {
+			return err
+		}
+
+		err = os.Remove(p.path)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func assembleParts(completedParts []*toDownload) (io.ReadCloser, error) {
+	p, err := createTempFile()
+	if err != nil {
+		return nil, err
+	}
+
+	err = appendPartsToFile(p, completedParts)
+	if err != nil {
+		return nil, err
+	}
+
+	f, err := os.Open(p)
+	if err != nil {
+		return nil, err
+	}
+
+	return &tempLocalObject{
+		path: p,
+		f:    f,
+	}, nil
+}
+
+func downloadParts(ctx context.Context, objectName, namespace, etag string, requestedSize, totalSize, maxBatchSize, offSet, minPartSize, maxPartNum int64, concurrentListeners int, downloadF downloadFunc) ([]*toDownload, error) {
+	numParts, partSize := getNumPartsAndPartSize(requestedSize, minPartSize, maxPartNum)
+	completedPartChan := make(chan *toDownload, numParts)
+	completedParts := make([]*toDownload, numParts)
+
+	batch := make([]*toDownload, 0)
```

Comment:
I don't think we need much if any channels here. Given the suggestion on OffsetWriter, this should be something like:

```go
f, err := os.Create(file)
if err != nil {
  return err
}
defer f.Close()
eg, egCtx := errgroup.WithContext(ctx)
eg.SetLimit(MaxConcurrency)
for i := 0; i < totalSize && egCtx.Err() == nil; i+=batchSize {
  wr := io.NewOffsetWriter(f, i)
  blobrange := clampedBlobRange(i, batchSize, totalSize)
  eg.Go(func() error {
    return downloadPart(egCtx, wr, objectName, namespace, etag, blobrange)
  })
}
return eg.Wait()
```

Reply:
theres something im not understanding about how to implement this correctly... even after writing tests that i think correctly copy data in parts correctly, somethings not quite right, im always getting the `invalid or corrupt tablefile` error...